### PR TITLE
UPC/API: Adding ucp_worker_wait_mem

### PR DESCRIPTION
**Pull request by shamisp**
Blocking API for waiting on memory update

Signed-off-by: Pavel Shamis (Pasha) <pasharesearch@gmail.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2926/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/966/ for details.

**Comment by yosefe** [comment]
@shamisp pls also fix the commit message UPC -> UCP/API

**Comment by shamisp** [comment]
sure

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2928/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/968/ for details.

**Comment by shamisp** [comment]
Notes from the call:
1. ucp_worker_wait() can be extended to use the addr *
2. Pasha: to verify the impact of inlining vs function call

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1164/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1165/ for details.

**Comment by shamisp** [comment]
The code demonstrates ~43% reduction in cycle count and it was tested with OSHMEM

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3092/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3093/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@yosefe please take a look now

**Comment by shamisp** [comment]
Link to openshmem test code https://github.com/shamisp/ompi/commit/a0e0fe9d54b61c017729a8922086143311154870

**Comment by yosefe** [comment]
@bbenton pls review as well

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -1127,6 +1128,25 @@ ucs_status_t ucp_worker_wait(ucp_worker_h worker);
 
 /**
  * @ingroup UCP_WAKEUP
+ * @brief Wait for memory update on the address
+ *
+ * This routine waits for a memory update at the local memory @a address.  This
+ * is a blocking routine. The routine returns when the memory address is
+ * updated ("write") or an event occurs in the system.
+ *
+ * @note This routine can be used by applications that "polls" for a memory
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: "polls" -> "poll"
or
applications -> an application

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -1127,6 +1128,25 @@ ucs_status_t ucp_worker_wait(ucp_worker_h worker);
...
+ * update. Instead of continues polling on an address the application may
```
Comment (bbenton) [inline src/ucp/api/ucp.h]: continues -> continuous

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1217/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3167/ for details (Mellanox internal link).

**Comment by bbenton** [comment]
:+1: for API/doc

**Pull request merged**